File: src/agent/agent-sdk/nodejs/examples/basic.ts
/*
 * 可运行示例：实现 AgentHandler，并用 DefaultBuildRunner 拉起真实 worker / docker 构建。
 *
 * 运行前请准备好后台网关与鉴权信息。两种方式提供配置：
 *   1) AgentConfig.fromPropertiesFile('/path/to/.agent.properties')
 *   2) 直接 new AgentConfig({...})（本示例采用）
 *
 * 用 ts-node 运行：  npx ts-node examples/basic.ts
 */

import * as os from 'os';
import * as path from 'path';
import {
  AgentConfig,
  AgentHandler,
  AgentLoop,
  BuildJobType,
  HeartExtra,
  HeartbeatContext,
  ImageDebug,
  StartupInfo,
  ThirdPartyBuildInfo,
  UpgradeItem,
  AgentHeartbeatResponse,
  registry,
  AgentApi,
  HttpClient,
  DefaultBuildRunner,
  downloadWorkerJar,
  downloadDockerInitFile,
  LogType,
} from '../src';

const AGENT_VERSION = '1.0.0-nodejs-sdk';
const WORKER_VERSION = '';
// ── Agent配置（TODO: 按你的环境填写）──────────────────────────────
const GATEWAY = ''; // TODO: 填写后台网关地址
const TOKEN = ''; // TODO: 填写 token/deviceId/userId 鉴权相关
const DEVICE_ID = ''; // TODO: 填写 token/deviceId/userId  鉴权相关
const USER_ID = ''; // TODO: 填写 token/deviceId/userId  鉴权相关
// ── 实机路径配置（TODO: 按你的环境填写）──────────────────────────────
const WORK_DIR = process.cwd(); // TODO: 填写 agent 工作目录
const WORKER_JAR_PATH = path.join(WORK_DIR, 'worker-agent.jar'); // TODO: 填写 worker-agent.jar 路径
const JDK17_PATH = ''; // TODO: 填写 jdk17 的 java 可执行文件或 JDK 目录
const JDK8_PATH = ''; // TODO: 填写 jdk8 的 java 可执行文件或 JDK 目录（可选）
const JDK17_DIR_PATH = ''; // TODO: docker 构建挂载用的 jdk17 目录（可选）
const JDK8_DIR_PATH = ''; // TODO: docker 构建挂载用的 jdk8 目录（可选）
const DOCKER_INIT_SCRIPT_PATH = path.join(WORK_DIR, 'agent_docker_init.sh'); // TODO: docker init 脚本路径
const ENABLE_DOCKER_BUILD = false; // TODO: 是否启用 docker 构建
const PARALLEL_TASK_COUNT = 4;
const DOCKER_PARALLEL_TASK_COUNT = 4;

/**
 * Handler：把构建动作委托给 DefaultBuildRunner，
 * 其它生命周期/心跳/状态查询仍由本类实现。
 */
class DemoHandler implements AgentHandler {
  private upgrading = false;

  constructor(private readonly runner: DefaultBuildRunner) { }

  onStartup(): StartupInfo {
    return {
      hostName: os.hostname(),
      hostIp: firstNonLoopbackIp(),
      detectOS: `${os.platform()}_${os.release()}`,
      masterVersion: AGENT_VERSION,
      version: WORKER_VERSION,
    };
  }

  collectHeartExtra(_ctx: HeartbeatContext, upgradeEnable: boolean): HeartExtra {
    return {
      masterVersion: AGENT_VERSION,
      slaveVersion: WORKER_VERSION,
      hostName: os.hostname(),
      agentIp: firstNonLoopbackIp(),
      agentInstallPath: WORK_DIR,
      startedUser: os.userInfo().username,
      props: {
        arch: os.arch(),
        jdkVersion: [],
        dockerInitFileMd5: { fileMd5: '', needUpgrade: false },
        osVersion: os.release(),
      },
      // 由 runner 维护正在运行的任务列表，供心跳上报
      override: {
        taskList: this.runner.getTaskList(),
        dockerTaskList: this.runner.getDockerTaskList(),
      },
      upgrade: upgradeEnable
        ? {
          workerVersion: WORKER_VERSION,
          goAgentVersion: AGENT_VERSION,
          jdkVersion: [],
          dockerInitFileInfo: { fileMd5: '', needUpgrade: false },
        }
        : null,
    };
  }

  isUpgrading(): boolean {
    return this.upgrading;
  }

  checkParallelTaskCount(): [boolean, boolean] {
    return this.runner.checkParallelTaskCount();
  }

  hasRunningJob(): boolean {
    return this.runner.hasRunningJob();
  }

  pipelineEnabled(): boolean {
    return false;
  }

  dockerDebugEnabled(): boolean {
    return false;
  }

  /** 领取到构建任务：直接委托给 runner，它会分发 worker/docker 并上报 finish。 */
  async onBuild(build: ThirdPartyBuildInfo): Promise<void> {
    console.log(`[demo] 领取到构建任务 buildId=${build.buildId} vmSeqId=${build.vmSeqId}`);
    await this.runner.onBuild(build);
    console.log(`[demo] 构建结束 buildId=${build.buildId}`);
  }

  onUpgrade(upgrade: UpgradeItem, hasBuild: boolean): void {
    console.log('[demo] 收到升级指令', upgrade, 'hasBuild=', hasBuild);
  }

  onPipeline(pipeline: Record<string, unknown>): void {
    console.log('[demo] 收到流水线指令', pipeline);
  }

  onImageDebug(debug: ImageDebug): void {
    console.log('[demo] 收到镜像调试指令 debugId=', debug.debugId);
  }

  onHeartbeatResp(resp: AgentHeartbeatResponse): void {
    console.log('[demo] 心跳响应，parallelTaskCount=', resp.parallelTaskCount);
  }

  onAgentDeleted(): void {
    console.warn('[demo] Agent 已被后台删除');
  }
}

function firstNonLoopbackIp(): string {
  const ifaces = os.networkInterfaces();
  for (const name of Object.keys(ifaces)) {
    for (const ni of ifaces[name] ?? []) {
      if (ni.family === 'IPv4' && !ni.internal) return ni.address;
    }
  }
  return '127.0.0.1';
}

async function main(): Promise<void> {
  const regisResp = await registry(GATEWAY, { token: TOKEN, deviceId: DEVICE_ID, userId: USER_ID });
  const config = new AgentConfig({
    gateway: GATEWAY,
    projectId: regisResp.projectId,
    agentId: regisResp.agentId,
    secretKey: regisResp.secretKey,
    fileGateway: regisResp.fileGateway,
    parallelTaskCount: regisResp.parallelTaskCount ?? PARALLEL_TASK_COUNT,
    dockerParallelTaskCount: regisResp.dockerParallelTaskCount ?? DOCKER_PARALLEL_TASK_COUNT,
    enableDockerBuild: ENABLE_DOCKER_BUILD,
  });

  // AgentApi 可独立构造（无需先建 AgentLoop），供 runner 上报 workerBuildFinish / 日志
  const client = new HttpClient({ timeoutMs: config.timeoutSec * 1000 });
  const api = new AgentApi(config, client);

  // 构造 runner：把日志转发到后台构建日志接口
  const runner = new DefaultBuildRunner({
    api,
    workDir: WORK_DIR,
    workerJarPath: WORKER_JAR_PATH,
    jdk17Path: JDK17_PATH,
    jdk8Path: JDK8_PATH,
    jdk17DirPath: JDK17_DIR_PATH,
    jdk8DirPath: JDK8_DIR_PATH,
    dockerInitScriptPath: DOCKER_INIT_SCRIPT_PATH,
    gateway: config.getGateway(),
    fileGateway: config.fileGateway,
    projectId: config.projectId,
    agentVersion: AGENT_VERSION,
    workerVersion: WORKER_VERSION,
    language: config.language,
    parallelTaskCount: config.parallelTaskCount,
    dockerParallelTaskCount: config.dockerParallelTaskCount,
    // 把 runner/docker 日志上报到后台（可选）
    postLog: (build, message) => {
      void api
        .addLogLine(
          build.buildId,
          {
            message,
            timestamp: Date.now(),
            tag: 'startVM-' + build.vmSeqId,
            jobId: build.containerHashId ?? '',
            logType: LogType.Log,
            executeCount: build.executeCount,
          },
          build.vmSeqId
        )
        .catch((e) => console.error('[demo] postLog 失败', e));
    },
    logFn: (msg) => console.log('[demo][runner]', msg),
  });

  // 用带 runner 的 handler 构造 loop
  const loop = new AgentLoop({
    config,
    handler: new DemoHandler(runner),
    intervalMs: 5000,
    monitorFn: () => console.log('[demo] 采集监控指标...'),
    monitorIntervalMs: 60_000,
    // 可选：启动时准备好 worker.jar / docker init 脚本
    onInit: async () => {
      const auth = config.getAuthHeaderMap();
      try {
        const r = await downloadWorkerJar(config.getGateway(), auth, WORK_DIR);
        console.log('[demo] worker.jar 就绪 md5=', r.md5, 'notModified=', r.notModified);
      } catch (e) {
        console.error('[demo] 下载 worker.jar 失败', e);
      }
      if (ENABLE_DOCKER_BUILD) {
        try {
          const r = await downloadDockerInitFile(config.getGateway(), auth, WORK_DIR);
          console.log('[demo] docker init 脚本就绪 md5=', r.md5);
        } catch (e) {
          console.error('[demo] 下载 docker init 脚本失败', e);
        }
      }
    },
  });

  process.on('SIGINT', () => {
    console.log('收到 SIGINT，停止 agent 主循环');
    loop.stop();
  });

  await loop.run();
}

void main();

// 让 BuildJobType 被引用，避免示例中未使用告警（实际使用见 checkBuildType 逻辑）
void BuildJobType;

File: src/agent/agent-sdk/python/examples/basic.py
"""Minimal composition of AgentLoop, DefaultBuildRunner, and worker management."""

from __future__ import annotations

import asyncio
import getpass
import os
from pathlib import Path
import platform
import socket

from bkci_agent_sdk import (
    AgentApi,
    AgentConfig,
    AgentLoop,
    DefaultBuildRunner,
    DefaultWorkerJarManager,
    HeartExtra,
    HttpClient,
    StartupInfo,
)

AGENT_VERSION = "1.0.0-python-sdk"
WORK_DIR = Path.cwd()
JDK17_PATH = os.environ.get("BKCI_JDK17_PATH", "")


class DemoHandler:
    def __init__(self, runner, worker_manager, api):
        self.runner = runner
        self.worker_manager = worker_manager
        self.api = api
        self.upgrading = False

    def on_startup(self):
        return StartupInfo(
            host_name=socket.gethostname(),
            host_ip="127.0.0.1",
            detect_os=f"{platform.system()}_{platform.release()}",
            master_version=AGENT_VERSION,
            version=self.worker_manager.get_version(),
        )

    def collect_heart_extra(self, _context, upgrade_enable):
        worker_version = self.worker_manager.get_version()
        return HeartExtra(
            master_version=AGENT_VERSION,
            slave_version=worker_version,
            host_name=socket.gethostname(),
            agent_ip="127.0.0.1",
            agent_install_path=str(WORK_DIR),
            started_user=getpass.getuser(),
            props={
                "arch": platform.machine(),
                "jdkVersion": [],
                "dockerInitFileMd5": {"fileMd5": "", "needUpgrade": False},
                "osVersion": platform.release(),
            },
            override={
                "taskList": self.runner.get_task_list(),
                "dockerTaskList": self.runner.get_docker_task_list(),
            },
            upgrade=(
                {
                    "workerVersion": worker_version,
                    "goAgentVersion": AGENT_VERSION,
                    "jdkVersion": [],
                    "dockerInitFileInfo": {"fileMd5": "", "needUpgrade": False},
                }
                if upgrade_enable
                else None
            ),
        )

    def is_upgrading(self): return self.upgrading
    def check_parallel_task_count(self): return self.runner.check_parallel_task_count()
    def has_running_job(self): return self.runner.has_running_job()
    def pipeline_enabled(self): return False
    def docker_debug_enabled(self): return False

    async def on_build(self, build):
        await self.runner.on_build(build)

    async def on_upgrade(self, upgrade, has_build):
        if not upgrade.get("worker") or has_build or self.upgrading:
            return
        self.upgrading = True
        success = False
        try:
            await self.worker_manager.upgrade()
            success = True
        finally:
            self.upgrading = False
            await self.api.finish_upgrade(success)

    async def on_pipeline(self, pipeline): pass
    async def on_image_debug(self, debug): pass
    async def on_heartbeat_resp(self, response): pass
    async def on_agent_deleted(self): pass


async def main() -> None:
    config = AgentConfig.from_properties_file()
    api = AgentApi(config, HttpClient(timeout_seconds=config.timeout_sec))
    worker_manager = DefaultWorkerJarManager(
        gateway=config.get_gateway(),
        auth_headers=config.get_auth_header_map(),
        work_dir=WORK_DIR,
        jdk17_path=JDK17_PATH,
    )
    await worker_manager.initialize()
    runner = DefaultBuildRunner(
        api=api,
        work_dir=WORK_DIR,
        worker_jar_path=worker_manager.get_worker_jar_path(),
        jdk17_path=JDK17_PATH,
        gateway=config.get_gateway(),
        file_gateway=config.file_gateway,
        project_id=config.project_id,
        agent_id=config.agent_id,
        secret_key=config.secret_key,
        agent_version=AGENT_VERSION,
        worker_version=worker_manager.get_version,
        parallel_task_count=config.parallel_task_count,
        docker_parallel_task_count=config.docker_parallel_task_count,
    )
    loop = AgentLoop(
        config=config,
        handler=DemoHandler(runner, worker_manager, api),
    )
    await loop.run()


if __name__ == "__main__":
    asyncio.run(main())

